sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2578,6 +2578,8 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
         executeSameContext(e.plan)
       }
 
+      checkAnalysis(newSubqueryPlan)
```

Comment:
This change is risky as we may fail earlier than before, while before we can still resolve the subquery expression after more iterations.

Reply:
Hmm, makes sense. Maybe restoring to the fix of moving `ResolveEncodersInUDF`?

Reply:
Okay. I will make the change (and maybe adjust the test).

Reply:
Btw, for the MergInto issue we encountered, because the ScalaUDF is put in a subquery plan by rewriting MergeInto rule. So moving `ResolveEncodersInUDF` before rewriting can fix the issue.

But it doesn't fix the general issue if ScalaUDF is in other subquery plan.

Reply:
I tend to fix the general issue instead of just fixing the corner issue we encountered.

Reply:
Let me fix the issue we encounter first. We can consider the general issue later.